src/groups/mqb/mqbc/mqbc_storageutil.cpp

```diff
@@ -2701,50 +2701,155 @@ void StorageUtil::forceFlushFileStores(FileStores* fileStores)
+                       purgedQueuesResultsVec,
+    bslmt::Latch*      latch,
+    int                partitionId,
+    StorageSpMapVec*   storageMapVec,
+    bslmt::Mutex*      storagesLock,
+    const FileStores*  fileStores,
+    const bsl::string& domainName)
 {
     // executed by *QUEUE_DISPATCHER* thread with the specified 'partitionId'
 
     // PRECONDITIONS
-    BSLS_ASSERT_SAFE(fileStores[partitionId]->inDispatcherThread());
+    BSLS_ASSERT_SAFE(fileStores);
+    BSLS_ASSERT_SAFE((*fileStores)[partitionId]->inDispatcherThread());
     BSLS_ASSERT_SAFE(latch);
-    BSLS_ASSERT_SAFE(purgedQueuesVec);
+    BSLS_ASSERT_SAFE(purgedQueuesResultsVec);
     BSLS_ASSERT_SAFE(0 <= partitionId);
     BSLS_ASSERT_SAFE(static_cast<unsigned int>(partitionId) <
-                     fileStores.size());
-    BSLS_ASSERT_SAFE(purgedQueuesVec->size() == fileStores.size());
+                     fileStores->size());
+    BSLS_ASSERT_SAFE(purgedQueuesResultsVec->size() == fileStores->size());
+    BSLS_ASSERT_SAFE(storageMapVec);
+    BSLS_ASSERT_SAFE(storagesLock);
 
-    bsl::vector<mqbcmd::PurgeQueueResult>& purgedQueues =
-        (*purgedQueuesVec)[partitionId];
-    fileStores[partitionId]->purgeDomain(&purgedQueues, domainName);
+    const mqbs::FileStore* fileStore = (*fileStores)[partitionId].get();
+
+    mqbs::StorageCollectionUtil::StorageFilter filter =
+        mqbs::StorageCollectionUtilFilterFactory::byDomain(domainName);
+
+    bsl::vector<mqbi::Storage*> domainStorages;
+    {
+        const StorageSpMap& partitionStorages = (*storageMapVec)[partitionId];
+
+        bslmt::LockGuard<bslmt::Mutex> guard(storagesLock);  // LOCK
+        for (StorageSpMap::const_iterator it = partitionStorages.cbegin();
+             it != partitionStorages.cend();
+             it++) {
+            if (filter(it->second.get())) {
+                domainStorages.push_back(it->second.get());
+            }
+        }
+        // Prepare a vector of storages to purge and release `storagesLock`
+        // as fast as possible.
+    }
+
+    bsl::vector<mqbcmd::PurgeQueueResult>& purgedQueuesResults =
+        (*purgedQueuesResultsVec)[partitionId];
+    purgedQueuesResults.reserve(domainStorages.size());
+
+    for (size_t i = 0; i < domainStorages.size(); i++) {
+        mqbcmd::PurgeQueueResult result;
+        // No need to pass a Semaphore here because we call it in
+        // a synchronous way
+        purgeQueueDispatched(&result, NULL, domainStorages[i], fileStore, "");
+        purgedQueuesResults.push_back(result);
+    }
 
     latch->arrive();
 }
 
 void StorageUtil::purgeQueueDispatched(
-    mqbcmd::PurgeQueueResult* purgedQueue,
+    mqbcmd::PurgeQueueResult* purgedQueueResult,
     bslmt::Semaphore*         purgeFinishedSemaphore,
-    mqbs::FileStore*          fileStore,
     mqbi::Storage*            storage,
-    const bsl::string*        appId)
+    const mqbs::FileStore*    fileStore,
```

Comment:
`const mqbs::FileStore*    fileStore` -> `const mqbs::FileStore&    fileStore`

Reply:
Not good to do this, because this procedure is intended to be used as a dispatched callback. We call this procedure from another thread, while copy-passing arguments. Even const-ref string here is being copied, so I would prefer to roll back it to string pointer again, since we have guaranteed that the string will live longer than the procedure's call.

Reply:
Yes, you are correct. Copy-passing pointer is better than copying the entire object. For `appId`, the string size will be smaller so it's not a big deal, but feel free to change.